HARK/core.py

```diff
@@ -594,10 +606,46 @@ def make_shock_history(self):
             (self.T_sim, self.AgentCount), dtype=bool
         )
 
+        # Also make blank arrays for the draws of newborns' initial conditions
+        for var_name in self.state_vars:
+            self.newborn_init_history[var_name] = (
+                np.zeros((self.T_sim, self.AgentCount)) + np.nan
+            )
+
+        # Record the initial condition of the newborns created by
+        # initialize_sim -> sim_births
+        for var_name in self.state_vars:
+            # Check whether the state is array-like or an aggregate
+            if hasattr(self.state_now[var_name], "shape"):
+                self.newborn_init_history[var_name][self.t_sim] = self.state_now[
+                    var_name
+                ]
+            else:
+                self.newborn_init_history[var_name][self.t_sim] = self.state_now[
+                    var_name
+                ]
+
         # Make and store the history of shocks for each period
         for t in range(self.T_sim):
+
+            # Deaths
             self.get_mortality()
             self.shock_history["who_dies"][t, :] = self.who_dies
+
+            # Initial conditions of newborns
+            if np.sum(self.who_dies) > 0:
+                for var_name in self.state_vars:
+                    # Check whether the state is array-like or an aggregate
+                    if hasattr(self.state_now[var_name], "shape"):
+                        self.newborn_init_history[var_name][
+                            self.t_sim, self.who_dies
+                        ] = self.state_now[var_name][self.who_dies]
+                    else:
+                        self.newborn_init_history[var_name][
+                            self.t_sim, self.who_dies
+                        ] = self.state_now[var_name]
```

Comment:
This extra machinery seems to be for the case where a model is keeping an 'aggregate' value is being tracked by the agent.

My understanding is that this is done with some things, like PermAggShock, but that this is questionable design, because really that could be seen as 'outside' the agent.

Can you describe a use case for when a newborn needs an aggregate value assigned to it at birth?

Reply:
Here's a use case:  One scenario in which financial market uncertainty can be a big deal is if aggregate financial risk is correlated with individual labor income risk.  Like, unemployment risk is high in a period when there has been a big rate-of-return shock.  That would require agents to know what has happened to the aggregate shock.  If we had a richly and carefully designed structure in which it was always going to be easy and obvious how to query the parent "economy" object in which an agent was embedded in a given period to find out what the shocks were, then maybe we wouldn't need to store them.  But it is definitely easier on the programmer to just be able to retrieve the info directly from the agent directly.

Reply:
Ok, thank you. That's a very clear use case.

What I'm trying to understand now is how `make_shock_history` is being changed in this PR to support that use case.

`make_shock_history` is called by the user in order to specify this pre-loaded shock behavior. The first thing it does is call `initialize_sim`. `initialize_sim` is being changed to update the state variables with the first line of the newborn initial history:

`self.state_now[var_name] = self.newborn_init_history[var_name][0]`

Then `make_shock_history`, a little above this line, changes the `newborn_init_history` data for time step `self.t_sim`. Since all this is being called just after the simulation is initialized, I think that `self.t_sim = 0` here. So, it's just copying the data back from `state_now` back into `newborn_init_history`.

Then in this last block... it's looping through all the ('future") time steps of the simulation, updating the `newborn_init_history` data, but now drawing from the `state_now` data, which is the timestep 0 data for all agents.

So I think what's going to happen is that after this code is run, `newborn_init_history` will, for agent X at time step 100 (for example), have the same newborn state values as agent X at time step 0. This would be true for both the idiosyncratic agent state and the aggregate state.

So I'm quite confused -- why would agent X, reborn at time 100, have the aggregate state from time 0, when agent Y, who is not reborn then, would have a different aggregate state (that from time 100).

Am I missing something?

Reply:
> This extra machinery seems to be for the case where a model is keeping an 'aggregate' value is being tracked by the agent.
> 
> My understanding is that this is done with some things, like PermAggShock, but that this is questionable design, because really that could be seen as 'outside' the agent.
> 
> Can you describe a use case for when a newborn needs an aggregate value assigned to it at birth?

I think Chris's example above works.

But for my own simpler example with the current classes that are there:

Imagine an `IndShock` agent type with aggregate growth. Newborns' `pLvl`s are initialized taking the aggregate `pLvl` into account. It is therefore important to keep track of it, because it is needed to initialize the newborns.

Where should it be stored? Not sure. I agree it is not a property of any individual agent. But the actual agent type is in charge of tracking and updating it, even if there is no Economy class. That's why I am storing it there.

Reply:
> So I'm quite confused -- why would agent X, reborn at time 100, have the aggregate state from time 0, when agent Y, who is not reborn then, would have a different aggregate state (that from time 100).
> 
> Am I missing something?

You are right, its a bug. It should be `t` (the variable changing in the for loop), not `self.t_sim`

Reply:
Wait, nevermind. `Agent.t_sim` is being advanced inside the loop
https://github.com/Mv77/HARK/blob/6c615d20a062af41c22a8b14de33f05655c3f26f/HARK/core.py#L653
So I believe `t_sim` must track `t`. Not sure why I am updating `t_sim` and not using `t`. Definitively merits a review on my end.

Thank you!

Reply:
@sbenthall in any case, the (poorly executed) idea is the following.

Before running a full simulation, I want to
a) Draw all the mortality shocks for the simulation.
b) Draw the states that the newborns will get when they die.
c) Store those states.

I achieve that by creating arrays full of `nan`, one for every state. Then running a for through the "future" simulation steps. At every step I:
- Drawing mortality shocks and store them.
- Draw states for agents who die, using `get_mortality`, which puts those states in `state_now`.
- Copying those states *only for agents who died* into my `nan` arrays.